### ISSUE TITLE
Add ability to scrape external html

### ISSUE THREAD

**Issue by oveddan** [medialab/artoo#222]
Lets say we have a script that downloads external html, and we want to scrape it.  It would be great if artoo could be used for this, instead of always scraping the current page.

**Comment by Yomguithereal** [comment]
Hello @oveddan,
actually you can already do that. The `artoo.scrape` method can be given a jQuery selector rather than a string describing the selection so you can apply to other parsed html.

There is even a little helper called [`artoo.helpers.jquerify`](http://medialab.github.io/artoo/helpers/#jquerify) that will parse a lot of different HTML/XML strings without having to feed them to the DOM and avoid silly bugs like loading images etc.

Alternatively you can use ` $(sel).scrape` to do the job where $ points to another document.

*Example*

```js
var htmlString = '<div>Hey</div>';

// Parsing the html with caution (or through your own way if you prefer)
var $doc = artoo.helpers.jquerify(htmlString);

// Then scrape likewise
var data = artoo.scrape($doc.find('div'), ...});

// or
var data = $doc.find('div').scrape(...);
```

As a side note, the [ajax spider](http://medialab.github.io/artoo/spiders/) already have some sugar to fetch parse and scrape the targeted url's html.

**Comment by Yomguithereal** [comment]
You can even use artoo with node.js and cheerio if needed.

**Comment by oveddan** [comment]
I can't include  jQuery in my project - is there a way to do it without jQuery?

**Comment by Yomguithereal** [comment]
No, currently jQuery is a dependency of artoo so if you remove jQuery, it won't work. It *might* work with subsets of kind of subsets like Zepto etc. But, why can't you include jQuery in your project? Do you have another variable named `$`?

**Comment by oveddan** [comment]
Compatibility issues with `react-native`.  I was hoping for a simple lightweight dom parsing library independent of jquery.

**Comment by Yomguithereal** [comment]
Ok, I understand your use case better now. Do you have a precise library in mind to perform the parsing? My aim, lately, was to rebuild the library on at least commonJS so I could clearly define and require different parts of the library. I guess that what you are interested in is `artoo.scrape`.

I could fiddle something to see whether it would be easy to hook the scraping utilities on some other parsing library.

**Comment by oveddan** [comment]
Sadly, jquery is the best parsing library I know!  So I can see why it's a natural fit for artoo.  There is the html5 [DOMParser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser)  But all that does is convert html into a dom document.

**Comment by oveddan** [comment]
Would artoo.scrape work without jquery?

**Comment by oveddan** [comment]
Also possible to leverage [querySelector](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector)

**Comment by Yomguithereal** [comment]
For the time being, `artoo.scrape` wouldn't work without jQuery. But I guess it could be possible to fiddle something to leverage querySelector (while of course dropping all of the jQuery sugar in the process).

Other options could be to use  `Sizzle` (the jQuery parser) alone, or even see whether cheerio could work in a browser env (but I assume you are not strictly in a browser env in your use case). Btw, what are the restrictions imposed by `react-native` concerning the libraries you may use?

**Comment by Yomguithereal** [comment]
Note that if cheerio can be used with `react-native` then you can use the node version of artoo in your case.

**Comment by oveddan** [comment]
The main restriction is actually in `Sizzle` - it attempts to call `document.createElement`, but that's not supported in `react-native`.

**Comment by Yomguithereal** [comment]
Ok. So I guess you should try cheerio with artoo's [node](http://medialab.github.io/artoo/node/) version then. It does not rely on the DOM whatsoever and is actually faster than Sizzle.

**Comment by Yomguithereal** [comment]
Any update about this @oveddan?

**Comment by oveddan** [comment]
Thanks for following up @Yomguithereal

I actually got jQuery working within jscore (which is what react-native uses) by building jQuery and modifying one -https://github.com/jquery/jquery/blob/66e1b6b8d49812239b5712d65922ff94c60f7b02/src/intro.js#L25
and removing `global.document ?`, because with jscore there is a global document, but it does not have createElement which is what sizzle is looking for and breaks when it doesn't exist.

I think artoo would almost get me there - the challenge it seems is that with artoo in node you do:

```javascript
var $ = cheerio.load(myXMLString);
// Setting artoo's context
artoo.setContext($);
```

This essentially binds the global context of artoo to one document - when I could be scraping multiple documents concurrently.

It would be better if it worked like:

```javascript
var $ = cheerio.load(myXMLString);
// Setting artoo's context:
var context = artoo.Context($);
var data = context.scrape('ul > li', params);
```

Just as with jQuery in node - you do:
```
var $ = require('jquery')(window);
```

**Comment by Yomguithereal** [comment]
Hello @oveddan.
For the case where you want to scrape from multiple contexts, I usually prefer the following:

```js
var artoo = require('artoo-js'),
    cheerio = require('cheerio');

artoo.bootstrap(cheerio);

var $ = cheerio.load('whatever');

// Then either
var data = $(sel).scrape({...});
// or
var data = artoo.scrape($(sel), {...});
```
